accounts/pkg/storage/cs3.go

```diff
@@ -0,0 +1,293 @@
+package storage
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"path"
+	"strings"
+
+	user "github.com/cs3org/go-cs3apis/cs3/identity/user/v1beta1"
+	v1beta11 "github.com/cs3org/go-cs3apis/cs3/rpc/v1beta1"
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	"github.com/cs3org/reva/pkg/rgrpc/todo/pool"
+	"github.com/cs3org/reva/pkg/token"
+	"github.com/cs3org/reva/pkg/token/manager/jwt"
+	"github.com/owncloud/ocis/accounts/pkg/config"
+	"github.com/owncloud/ocis/accounts/pkg/proto/v0"
+	"google.golang.org/grpc/metadata"
+)
+
+// CS3Repo provides a cs3 implementation of the Repo interface
+type CS3Repo struct {
+	serviceID       string
+	cfg             *config.Config
+	tm              token.Manager
+	storageProvider provider.ProviderAPIClient
+	dataProvider    dataProviderClient // Used to create and download data via http, bypassing reva upload protocol
+}
+
+// NewCS3Repo creates a new cs3 repo
+func NewCS3Repo(serviceID string, cfg *config.Config) (Repo, error) {
+	tokenManager, err := jwt.New(map[string]interface{}{
+		"secret": cfg.TokenManager.JWTSecret,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := pool.GetStorageProviderServiceClient(cfg.Repo.CS3.ProviderAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	return CS3Repo{
+		serviceID:       serviceID,
+		cfg:             cfg,
+		tm:              tokenManager,
+		storageProvider: client,
+		dataProvider: dataProviderClient{
+			client: http.Client{
+				Transport: http.DefaultTransport,
+			},
+		},
+	}, nil
+}
+
+// WriteAccount writes an account via cs3 and modifies the provided account (e.g. with a generated id).
+func (r CS3Repo) WriteAccount(ctx context.Context, a *proto.Account) (err error) {
+	t, err := r.authenticate(ctx)
+	if err != nil {
+		return err
+	}
+
+	ctx = metadata.AppendToOutgoingContext(ctx, token.TokenHeader, t)
+	if err := r.makeRootDirIfNotExist(ctx, accountsFolder); err != nil {
+		return err
+	}
+
+	var by []byte
+	if by, err = json.Marshal(a); err != nil {
+		return err
+	}
+
+	_, err = r.dataProvider.put(r.accountURL(a.Id), bytes.NewReader(by), t)
+	return err
+}
+
+// LoadAccount loads an account via cs3 by id and writes it to the provided account
+func (r CS3Repo) LoadAccount(ctx context.Context, id string, a *proto.Account) (err error) {
+	t, err := r.authenticate(ctx)
+	if err != nil {
+		return err
+	}
+
+	resp, err := r.dataProvider.get(r.accountURL(id), t)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode == http.StatusNotFound {
+		return &notFoundErr{"account", id}
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
```

Comment:
I would not use `ioutil.ReadAll` as it loads everything in memory. Instead I would use a buffered decoder. Something like this:

```golang
package main

import (
	"encoding/json"
	"net/http"
	"log"
	"fmt"
)

func main() {
	resp, err := http.Get("https://api.mocki.io/v1/b043df5a")
	if err != nil {
		log.Fatal(err)
	}
	defer resp.Body.Close()

	var contents interface{}
	dec := json.NewDecoder(resp.Body)
	if err := dec.Decode(&contents); err != nil {
		log.Fatal(err)
	}
	fmt.Println(contents)
}

```

[Here is a good short read](https://haisum.github.io/2017/09/11/golang-ioutil-readall/)

Reply:
😍